### PR TITLE
[RealImage/apsara #921] Fix getting state in the end of playback

### PR DESCRIPTION
**Pull request by VValkovoy**
# What this PR does?
* Fixes playback of `gst-plugin-imbsink` Gstreamer plugin regarding getting valid state in the end of playback.

- **Issue**
 Currently ImbSink plugin hangs in the end of playback without giving any response to MS.

- **Solution**
    * Fix hanging state of gst-plugin-imbsink in the end of playback

- **Expected behavior**
`gst-plugin-imbsink` Gstreamer plugin should provide playback without hanging in the end woth valid get state response.

# How to build:
* `git clone --recursive -b feature/fix-freeze git@github.com:RealImage/gst-plugin-imbsink.git`
* `cmake  ..`
* `make -j4`

### CODE REVIEW

File: src/ImbRenderer.cpp [hunk 1]
```diff
@@ -230,6 +230,9 @@ bool ImbRenderer::WaitForPlaybackToComplete()
         last_buffer_rendered = true;
     }
 
+    _videoBufferQueue->StopBuffering();
```
Comment (Iyer-VivekAnand) [inline src/ImbRenderer.cpp]: I am finding it a little difficult to understand the flow. Can you come up with a sequence diagram?

File: src/ImbRenderer.cpp [hunk 2]
```diff
@@ -252,8 +255,8 @@ void ImbRenderer::_Play()
 
             /* Stop by EOS */
             /* No need to render buffers if EOS received and queues are empty */
-            if ((_videoBufferQueue->IsEos() || _audioBufferQueue->IsEos()) && !_videoBufferQueue->GetSize() &&
-                !_audioBufferQueue->GetSize())
+            if ((_videoBufferQueue->IsEos() || _audioBufferQueue->IsEos()) &&
+                (!_videoBufferQueue->GetSize() || !_audioBufferQueue->GetSize()))
```
Comment (AntonQubecinema) [inline src/ImbRenderer.cpp]: @VValkovoy  Can't we just ask IMB and get its state to decide that playback is finished?

File: src/BufferQueue.cpp [hunk 3]
```diff
@@ -58,15 +58,15 @@ bool MediaBlock::BufferQueue::Push(GstBuffer* buffer)
         while (queue_overflow)
         {
             GST_WARNING("Buffer Queue is overrun. Queue size is [%d]. Waiting to be popped ...", buffer_queue->size());
-            if (GetBufferStatus() == BufferingStatus::STOPPED)
+            if (GetBufferStatus() == BufferingStatus::STOPPED || GetBufferStatus() == BufferingStatus::EOS_REACHED)
```
Comment (AntonQubecinema) [inline src/BufferQueue.cpp]: Why we have both states here? please provide examples for both cases

File: src/ImbRenderer.cpp (hunk 2 above)
Comment (VValkovoy) [reply src/ImbRenderer.cpp]: No, @AntonQubecinema , we can`t. As for it is us who drive the playback control of IMB.
Currently we have synchronized `V+A` stream flowing from ImbSinkBin.
That means that when all N frames of Video & Audio reels have been received by ImbSink plugin and the last one has Delivered all pairs of V+A accumulated inside Queues to IMB, the next come processing EOS which waits until all buffers have been absorbed by Decoders from PreDecoder buffers aboard IMB.

File: src/ImbRenderer.cpp (hunk 1 above)
Comment (VValkovoy) [reply src/ImbRenderer.cpp]: @Iyer-VivekAnand 
Please refer to Sequence diagram of Imbsink plugin relating to Push/Pop & EOS processing.
https://github.com/RealImage/apsara/blob/198e13150e3e1ae8af3d2fa4bcb85e0b6a56470b/design/images/Imbsink_structure_workflow.png

File: src/BufferQueue.cpp (hunk 3 above)
Comment (VValkovoy) [reply src/BufferQueue.cpp]: Fixed onto allowing to Push only in Awaiting states.

**Pull request merged**